Correct to_string implementation for ObjectId.

jsantos17 opened the pull request:
Fixes #2974, Fixes #2969 

Not sure if including `Id` in `Syntaxed` makes sense. Opening this for discussion and Travis. `Syntaxed` is used only for `ToString`, so I don't think it'll lead to issues in the future or in other connectors.

jsantos17 commented on frontend/src/main/scala/quasar/types.scala:
```diff
@@ -531,7 +531,7 @@ object Type extends TypeInstances {
   val Temporal = Timestamp ⨿ Date ⨿ Time
   val Comparable = Numeric ⨿ Interval ⨿ Str ⨿ Temporal ⨿ Bool
-  val Syntaxed = Type.Null ⨿ Type.Comparable
+  val Syntaxed = Type.Null ⨿ Type.Comparable ⨿ Type.Id
```
This is necessary for `to_string(oid(_id))`. Not sure if that's something we want to allow. It's also necessary for `objectIdToString.test` since AFAIK there's no way to load an `ObjectId` into Mongo for `it` tests.

jsantos17 replied:
I guess converting an `Id` to `String` is something that might be useful in general, for any connector.

wemrysi replied:
The `Id` type is dubious and seems very mongo-centric, but as it already exists and exposed in SQL<sup>2</sup> I'll give a reluctant 👍. It will go away once we switch to the new type representation in favor of metadata.

@jsantos17 Since you're changing `Syntaxed` you'll need to update the typer/untyper for `StringLib.ToString` to handle `Id`.

jsantos17 commented on build.sbt:
```diff
@@ -38,6 +38,7 @@ lazy val buildSettings = commonBuildSettings ++ Seq(
 
   scalacOptions ++= Seq(
     "-target:jvm-1.8",
+    "-Ypatmat-exhaust-depth", "40",
```
`MongoDbPlanner` does not compile without this. Not sure why.

djspiewak replied:
Note that most of the pattern matching on qscript/lp is already not checked for exhaustiveness because of a bug in scalac.  You can thus get around exhaustiveness depth warnings by using `@unchecked` on the match target without actually changing anything about the safety of the patmat.

jsantos17 replied:
What bug is this? I'd like to add a comment above `@unchecked` linking it.

rintcius replied:
Any problems with adding `@unchecked` and removing `"-Ypatmat-exhaust-depth", "40"`?

rintcius commented on mongodb/src/main/scala/quasar/physical/mongodb/MongoDbPlanner.scala:
```diff
@@ -336,6 +344,7 @@ object MongoDbPlanner {
 
     val handleSpecialCore: MapFuncCore[T, JsCore] => M[JsCore] = {
       case Constant(v1) => ejsonToJs[M, T[EJson]](v1)
+      case ToId(a1) => New(Name("ObjectId"), List(a1)).point[M]
```
move to `JsFuncHandler` instead?

wemrysi commented on mongodb/src/main/scala/quasar/physical/mongodb/MongoDbPlanner.scala:
```diff
@@ -212,7 +212,15 @@ object MongoDbPlanner {
-      case ToString(a1) => unimplemented[M, Fix[ExprOp]]("ToString expression")
+      // NB: Using $substr or $substrBytes to convert to string does not work for ObjectId, so we force
+      //     this to be planned using JS
+      case ToString($var(DocVar(DocVar.ROOT, Some(BsonField.Name("_id"))))) =>
```
How about moving this to a new function that is run before the `funcHandler`? This would allow this special case to be intercepted and failed while preserving the versioned variants of `ToString`.

rintcius replied:
@wemrysi You mean with new function something to use instead of `ToString` (like `ObjectIdToString`), right?

wemrysi replied:
No, I mean a new scala function that is called before `funcHandler`. Right now we have `funcHandler` -> `handleSpecial`, I'm proposing `newFunction` -> `funcHandler` -> `handleSpecial`.

rintcius replied:
But how do you see this being intercepted just for ObjectId's then?

wemrysi replied:
We'd match on the case above and fail, but let everything else through, to be handled by the versioned `funcHandler`.

rintcius replied:
Ok, that addresses just the versioning issue then? 
A more serious issue is that the code does not properly recognize ObjectId's. We'd still have that issue, if I understand you correctly

wemrysi replied:
Yes I understand the overall approach fails in many cases, I'm just suggesting an improvement if this is the approach that ends up being taken.

jsantos17 replied:
Yes, I don't really see a way to keep the `ToString`  aggregation impl while supporting `ToString` on `ObjectId`s. 

There's nothing to tell us we're converting an `ObjectId`, so we rely on the `_id` string here. Not ideal, but I see no way around it. Hopefully someone else can chime in with other ideas.

jsantos17 commented:
@rintcius ready for another look, except for the versioned `ToString` since I'm still not sure what approach we're going to go with.

rintcius commented on frontend/src/main/scala/quasar/std/string.scala:
```diff
@@ -342,6 +342,7 @@ trait StringLib extends Library {
         case Data.Date(d)      => success(d.toString)
         case Data.Time(t)      => success(t.format(DataCodec.timeFormatter))
         case Data.Interval(i)  => success(i.toString)
+        case Data.Id(i)        => success(i.toString)
```
I think it's worth adding a test for this if it's not too difficult (I think using `oid(<some literal hex>)` would hit this case)

jsantos17 replied:
Yes, I was unable to use unchecked. I tried `(func: @unchecked) match [...]`, refactoring into a function to tag it with unchecked, but the same error still comes up. What's the correct way to do it?

jsantos17 replied:
Added a test for this.

rintcius commented on mongodb/src/main/scala/quasar/physical/mongodb/MongoDbPlanner.scala:
```diff
@@ -601,22 +485,12 @@ object MongoDbPlanner {
-            "map"),
-          List(
-            Fun(List(Name("element"), Name("index")),
-              BinOp(jscore.Add, ident("index"), start)))).point[M]
```
Needs an extra `case _ => unimplemented[M, JsCore]("JS function")` - would be nice if we could add the name of the specific mapFunc (like is done eg in `handleSpecialDerived`), but I think this is not implemented yet, so that could be added as a TODO

jsantos17 commented:
@rintcius I added `version: revision`. This PR doesn't change any APIs since the `oid` function was already exposed. I just implemented it in new Mongo.
